File: js/consume.js
const WebSocket = require('ws');
var parquet = require('parquetjs-lite');

const wss = new WebSocket('wss://stream.zash.sh', {
  headers: { 'x-api-key': 'YOUR_API_KEY' },
});

function process(trades) {
  // insert here your own custom processor
  console.log(trades);
}

wss.addEventListener('message', function (event) {
  const raw = Buffer.from(event.data, 'base64');
  read(raw).then(process).catch(console.error);
});

async function read(buffer) {
  let reader = await parquet.ParquetReader.openBuffer(buffer);
  let cursor = reader.getCursor();
  let record = null;
  let results = [];
  while ((record = await cursor.next())) {
    results.push(record);
  }
  return results;
}

File: python/consume.py
import io
import websockets
import asyncio
import pandas as pd
import base64


def process(trades):
    # insert your own trade processor here
    print(trades)


async def connect():
    async with websockets.connect(
        "wss://stream.zash.sh",
        extra_headers={"x-api-key": "YOUR_API_KEY"},
    ) as ws:
        while True:
            raw = await asyncio.wait_for(ws.recv(), timeout=180)
            process(
                pd.read_parquet(io.BytesIO(base64.b64decode(raw))).to_dict(
                    orient="records"
                )
            )


asyncio.run(connect())
